### DISCUSSION TITLE
replace the base editor

### DISCUSSION THREAD

**Discussion by Ahima49** [django-wiki/django-wiki#1211] [Q&A]
I use django-wiki app :
[https://django-wiki.readthedocs.io/en/main/][1]

I try to replace the base editor.
The default param for editor is :

    wiki.conf.settings.EDITOR = 'wiki.editors.markitup.MarkItUp'

I installed Martor (Markdown Editor plugin for Django) :
[https://github.com/agusmakmun/django-markdown-editor][2]

I don't know how properly load it.
I tried

    import martor
    wiki.conf.settings.EDITOR = 'martor'

and catch this error :

    ImportError: Could not import 'martor'. The path must be fully qualified.

I tried also many things and catch differents errors.

How properly change the django-wiki editors config ?
Thanks for reading.

ps: I use the last version of all packages (in a virtualenv)

    Linux Mint 21
    Python 3.10.4
    django 4.0.7
    Markdown==3.3.7
    martor==1.6.14
    wiki==0.9

**EDIT**
regarding the documentation of django-wiki :

> The following settings are available for configuration through your
> project. All settings are customized by prefixing WIKI_, so for
> instance URL_CASE_SENSITIVE should be configured as
> WIKI_URL_CASE_SENSITIVE. For plugins the prefix is WIKI_PLUGINNAME_,
> e.g. WIKI_IMAGES for the images plugin.

the param must be :

    WIKI_EDITOR = 'SOMETHING'

But the error steal the same with this way of import.

  [1]: https://django-wiki.readthedocs.io/en/main/
  [2]: https://github.com/agusmakmun/django-markdown-editor

**Comment by benjaoming** [comment]
Hi!

Great question!

To achieve what you want, you need to create your own subclass of `BaseEditor` like the existing one for `MarkItUp` and configure `WIKI_EDITOR` to its dotted path, i.e. `my_project.my_wiki_editor.Martor`

See: https://github.com/django-wiki/django-wiki/blob/main/src/wiki/editors/markitup.py

**Reply by Ahima49** [reply]
HI,
This is my own subclass of `BaseEditor`.

```python
from wiki.editors.base import BaseEditor
from django import forms

class MartorWidget(forms.Widget):
    template_name = "martor/bootstrap/editor.html"

    def __init__(self, attrs=None):
        # The 'rows' and 'cols' attributes are required for HTML correctness.
        default_attrs = {
            "class": "martor-preview",
            "rows": "10",
            "cols": "40",
        }
        if attrs:
            default_attrs.update(attrs)
        super().__init__(default_attrs)


class Martor(BaseEditor):
    editor_id = "martor"

    def get_widget(self, instance=None):
        return MartorWidget()

    class Media:
        css = {
            "all": (
                "martor/css/martor.bootstrap.css",
            )
        }
        js = (
            "martor/js/martor.bootstrap.js",
        )

```

I had configured the `WIKI_EDITOR` like that in settings.py

```python
WIKI_EDITOR = "wp_ahimawiki.Martor.Martor"
```
See the result

![post_git_issue_django-wiki](https://user-images.githubusercontent.com/32678765/191211413-fe06b7c7-23c3-45f8-846b-d48039bdad93.png)

If I extend the class MartorWidget instead form.Widget i load the editor but when i submit the form nothing are saved.

```python
from martor.widgets import `Widget`

class MartorWidget(MartorWidget):
    template_name = "martor/bootstrap/editor.html"
[...]
```
![screenshot_20092022-104809](https://user-images.githubusercontent.com/32678765/191212780-4034741c-26af-478c-83c8-e4a5e00a2ac5.png)

I know i don't understand all of the process..
Can you tell me what i have do wrong ?
Thanks

**Comment by benjaoming** [comment]
It looks like JS and CSS assets are loaded? Does the editor display as intended? Maybe you need to pay attention to the console and network traffic using your browser's Web Development mode? Have you read about how to configure Martor? Maybe you need to add extra configuration in JavaScript to make it work?
